File: scrape_reddit.py
import praw
import time
import json
from datetime import datetime

# Reddit API credentials
CLIENT_ID = "pUOCN3qZMewQhvtpuZo18w"
CLIENT_SECRET = "hDK6G7aclnN9ViDSIBtpHWHdlODRbg"
USERNAME = "EntrepreneurEdge"
PASSWORD = "Sasukekun@995835"
USER_AGENT = "ConvertScoutBot/1.0 by EntrepreneurEdge"

# Business details
TARGET_KEYWORD = "CRM"  # For leads
COMPETITORS = ["Salesforce"]  # For competitor complaints
COMPANY_NAME = "Hubspot"  # For complaints about your company

# Initialize PRAW
reddit = praw.Reddit(
    client_id=CLIENT_ID,
    client_secret=CLIENT_SECRET,
    username=USERNAME,
    password=PASSWORD,
    user_agent=USER_AGENT
)

# Function to check if a post/comment contains negative sentiment
def has_negative_sentiment(text):
    negative_keywords = [
        "lag", "slow", "buggy", "expensive", "overpriced", "terrible",
        "horrible", "frustrating", "bad", "issue", "problem", "delay",
        "sucks", "awful", "poor", "disappointing", "crap", "nightmare",
        "annoying", "useless", "broken", "clunky", "mess", "predatory",
        "unreliable", "confusing", "overhyped", "ridiculous", "waste",
        "scam", "disaster", "pathetic", "lame", "stupid", "dumb",
        "unintuitive", "bloated", "glitchy", "unresponsive", "trash"
    ]
    return any(keyword in text.lower() for keyword in negative_keywords)

# Function to fetch posts/comments from a subreddit
def fetch_reddit_data(query, limit, subreddits, require_negative_sentiment=False):
    results = []
    for subreddit_name in subreddits:
        try:
            subreddit = reddit.subreddit(subreddit_name)
            # Search for posts
            for submission in subreddit.search(query, limit=limit * 3):  # Fetch more to filter
                text = submission.title + " " + submission.selftext
                if require_negative_sentiment and not has_negative_sentiment(text):
                    continue
                results.append({
                    "username": submission.author.name if submission.author else "anonymous",
                    "platform": "Reddit",
                    "text": text,
                    "time": datetime.fromtimestamp(submission.created_utc).isoformat(),
                    "match": 90 + (len(results) * 2),  # Simulated match score (90–98%)
                    "connect_url": f"https://www.reddit.com{submission.permalink}"  # Post URL
                })
            # Search for comments
            for submission in subreddit.new(limit=15):  # Check more recent posts for comments
                submission.comments.replace_more(limit=0)
                for comment in submission.comments.list():
                    if query.lower() in comment.body.lower():
                        if require_negative_sentiment and not has_negative_sentiment(comment.body):
                            continue
                        results.append({
                            "username": comment.author.name if comment.author else "anonymous",
                            "platform": "Reddit",
                            "text": comment.body,
                            "time": datetime.fromtimestamp(comment.created_utc).isoformat(),
                            "match": 90 + (len(results) * 2),  # Simulated match score
                            "connect_url": f"https://www.reddit.com{comment.permalink}"  # Comment URL
                        })
        except Exception as e:
            print(f"Error fetching data from r/{subreddit_name}: {e}")
        if len(results) >= limit:
            break
    return results[:limit]

# Fetch data
def scrape_reddit_data():
    data = {
        "leads": [],
        "competitor_complaints": [],
        "company_complaints": []
    }

    # Subreddits to search (expanded list relevant to CRM, sales, tech, and more)
    subreddits = [
        "smallbusiness", "entrepreneur", "startups", "sales", "CRM",
        "SaaS", "marketing", "technology", "business", "techsupport",
        "sysadmin", "webdev", "programming", "software", "productivity",
        "growthhacking", "digitalmarketing", "B2B", "ecommerce", "freelance",
        "ITManagers", "smallbiz", "marketingstrategy", "businessowners", "tech"
    ]

    # Fetch leads (5 posts/comments, no negative sentiment required)
    print("Fetching leads from Reddit...")
    leads_query = TARGET_KEYWORD
    data["leads"] = fetch_reddit_data(leads_query, 5, subreddits, require_negative_sentiment=False)
    time.sleep(5)  # Delay to avoid rate limits

    # Fetch competitor complaints (2 posts/comments total, require negative sentiment)
    print("Fetching competitor complaints from Reddit...")
    competitor_complaints = []
    for competitor in COMPETITORS:
        complaint_query = f"{competitor} lag OR slow OR buggy OR expensive OR overpriced OR terrible OR horrible OR frustrating OR bad OR issue OR problem OR delay OR sucks OR awful OR poor OR disappointing OR crap OR nightmare OR annoying OR useless OR broken OR clunky OR mess OR predatory OR unreliable OR confusing OR overhyped OR ridiculous OR waste OR scam OR disaster OR pathetic OR lame OR stupid OR dumb OR unintuitive OR bloated OR glitchy OR unresponsive OR trash"
        complaints = fetch_reddit_data(complaint_query, 2, subreddits, require_negative_sentiment=True)
        competitor_complaints.extend(complaints)
        time.sleep(5)  # Delay to avoid rate limits
    data["competitor_complaints"] = competitor_complaints[:2]  # Limit to 2 total

    # Fetch complaints about your company (2 posts/comments, require negative sentiment)
    print("Fetching company complaints from Reddit...")
    company_complaint_query = f"{COMPANY_NAME} delay OR slow OR bad OR issue OR lag OR buggy OR expensive OR overpriced OR terrible OR horrible OR frustrating OR problem OR sucks OR awful OR poor OR disappointing OR crap OR nightmare OR annoying OR useless OR broken OR clunky OR mess OR predatory OR unreliable OR confusing OR overhyped OR ridiculous OR waste OR scam OR disaster OR pathetic OR lame OR stupid OR dumb OR unintuitive OR bloated OR glitchy OR unresponsive OR trash"
    data["company_complaints"] = fetch_reddit_data(company_complaint_query, 2, subreddits, require_negative_sentiment=True)
    time.sleep(5)  # Delay to avoid rate limits

    # Save data to a JSON file
    with open("reddit_data.json", "w") as f:
        json.dump(data, f, indent=4)
    print("Data saved to reddit_data.json")

if __name__ == "__main__":
    scrape_reddit_data()

File: scrape_reddit.js
const snoowrap = require("snoowrap");
require("dotenv").config();

const r = new snoowrap({
  userAgent: process.env.REDDIT_USER_AGENT,
  clientId: process.env.REDDIT_CLIENT_ID,
  clientSecret: process.env.REDDIT_CLIENT_SECRET,
  username: process.env.REDDIT_USERNAME,
  password: process.env.REDDIT_PASSWORD,
});

const delay = (ms) => new Promise((res) => setTimeout(res, ms));

const complaintTriggers = [
  "hate", "sucks", "terrible", "worst", "not working", "cancelled", "frustrated",
  "bug", "glitch", "billing issue", "broken", "scam", "avoid", "issue", "problem",
  "doesn't work", "unreliable", "disappointed", "too expensive", "support is bad",
  "not responsive", "slow", "crashing", "left", "switched"
];

const excludedTerms = [
  "job", "salary", "hiring", "resume", "intern", "school", "nsfw", "porn", "onlyfans"
];

function classify(text) {
  const t = text.toLowerCase();
  if (complaintTriggers.some(word => t.includes(word))) return "complaint";
  if (t.includes("looking for") || t.includes("recommend") || t.includes("alternative")) return "interest";
  return null;
}

function generateQueries({ problemSolved, competitor, businessName, targetCustomer, industryKeywords, painSummary }) {
  const base = [];

  const all = [
    ...(problemSolved ? [problemSolved] : []),
    ...(targetCustomer ? [targetCustomer] : []),
    ...(painSummary ? [painSummary] : []),
    ...(industryKeywords ? industryKeywords.split(",") : []),
  ];

  all.forEach(phrase => {
    const trimmed = phrase.trim().toLowerCase();
    base.push(`looking for ${trimmed}`);
    base.push(`recommendation for ${trimmed}`);
    base.push(`suggestions for ${trimmed}`);
    base.push(`tool to solve ${trimmed}`);
  });

  if (competitor) {
    const c = competitor.toLowerCase();
    base.push(`alternatives to ${c}`);
    base.push(`issues with ${c}`);
    base.push(`frustrated with ${c}`);
    base.push(`leaving ${c}`);
  }

  if (businessName) {
    const b = businessName.toLowerCase();
    base.push(`problems with ${b}`);
    base.push(`complaints about ${b}`);
    base.push(`experience using ${b}`);
  }

  return [...new Set(base)];
}

function getSubreddits() {
  return [
    "legaltech",
    "lawfirm",
    "smallbusiness",
    "startups",
    "Entrepreneur",
    "AskReddit",
    "SaaS"
  ];
}

const scrapeReddit = async (niche, competitor, businessName, problemSolved, targetCustomer = "", industryKeywords = "", painSummary = "") => {
  const subreddits = getSubreddits();
  const queries = generateQueries({ problemSolved, competitor, businessName, targetCustomer, industryKeywords, painSummary });

  const leads = [];
  const competitorComplaints = [];
  const companyComplaints = [];

  let throttleCounter = 0;

  for (const subreddit of subreddits) {
    for (const query of queries) {
      throttleCounter++;
      if (throttleCounter % 4 === 0) await delay(800);

      try {
        const posts = await r.getSubreddit(subreddit).search({
          query,
          sort: "relevance",
          time: "year",
          limit: 12,
        });

        for (const post of posts) {
          const raw = post.title + " " + (post.selftext || "");
          const lower = raw.toLowerCase();

          if (
            excludedTerms.some(term => lower.includes(term)) ||
            raw.length < 50 ||
            post.ups < 3
          ) continue;

          const type = classify(lower);
          if (!type) continue;

          let profilePicture = null;
          try {
            const user = await r.getUser(post.author.name).fetch();
            profilePicture = user.icon_img || null;
          } catch {}

          const postData = {
            username: post.author.name,
            platform: "Reddit",
            time: new Date(post.created_utc * 1000).toISOString(),
            text: raw,
            match: post.ups,
            connect_url: `https://reddit.com${post.permalink}`,
            profile_picture: profilePicture,
          };

          if (type === "interest") leads.push(postData);
          if (competitor && lower.includes(competitor.toLowerCase())) competitorComplaints.push(postData);
          if (businessName && lower.includes(businessName.toLowerCase())) companyComplaints.push(postData);
        }
      } catch (err) {
        console.log(`❌ Query "${query}" on r/${subreddit} failed:`, err.message);
      }
    }
  }

  return {
    leads: leads.slice(0, 3),
    competitorComplaints: competitorComplaints.slice(0, 3),
    companyComplaints: companyComplaints.slice(0, 3),
  };
};

module.exports = scrapeReddit;